source/common/upstream/maglev_lb.cc

```diff
@@ -54,26 +93,111 @@ MaglevTable::MaglevTable(const NormalizedHostWeightVector& normalized_host_weigh
+      table_(absl::bit_width(normalized_host_weights.size()), table_size) {
+  constructMaglevTableInternal(normalized_host_weights, max_normalized_weight,
+                               use_hostname_for_hashing);
+}
 
-  uint64_t min_entries_per_host = table_size_;
-  uint64_t max_entries_per_host = 0;
+void CompactMaglevTable::constructImplementationInternals(
+    std::vector<TableBuildEntry>& table_build_entries, double max_normalized_weight) {
+  // Populate the host table. Index into table_build_entries[i] will align with
+  // the host here.
+  host_table_.reserve(table_build_entries.size());
   for (const auto& entry : table_build_entries) {
-    min_entries_per_host = std::min(entry.count_, min_entries_per_host);
-    max_entries_per_host = std::max(entry.count_, max_entries_per_host);
+    host_table_.emplace_back(entry.host_);
   }
-  stats_.min_entries_per_host_.set(min_entries_per_host);
-  stats_.max_entries_per_host_.set(max_entries_per_host);
+  host_table_.shrink_to_fit();
 
-  if (ENVOY_LOG_CHECK_LEVEL(trace)) {
-    for (uint64_t i = 0; i < table_.size(); i++) {
-      const absl::string_view key_to_hash = hashKey(table_[i], use_hostname_for_hashing);
-      ENVOY_LOG(trace, "maglev: i={} address={} host={}", i, table_[i]->address()->asString(),
-                key_to_hash);
+  // Vector to track whether or not a given fixed width bit is set in the
+  // BitArray used as the maglev table.
+  std::vector<bool> occupied(table_size_, false);
+
+  // Iterate through the table build entries as many times as it takes to fill up the table.
+  uint64_t table_index = 0;
+  for (uint32_t iteration = 1; table_index < table_size_; ++iteration) {
+    for (uint32_t i = 0; i < table_build_entries.size() && table_index < table_size_; i++) {
+      TableBuildEntry& entry = table_build_entries[i];
+      // To understand how target_weight_ and weight_ are used below, consider a host with weight
+      // equal to max_normalized_weight. This would be picked on every single iteration. If it had
+      // weight equal to max_normalized_weight / 3, then it would only be picked every 3 iterations,
+      // etc.
+      if (iteration * entry.weight_ < entry.target_weight_) {
+        continue;
+      }
+      entry.target_weight_ += max_normalized_weight;
+      // As we're using the compact implementation, our table size is limited to
+      // 32-bit, hence static_cast here should be safe.
+      uint32_t c = static_cast<uint32_t>(permutation(entry));
+      while (occupied[c]) {
+        entry.next_++;
+        c = static_cast<uint32_t>(permutation(entry));
+      }
+
+      // Record the index of the given host.
+      table_.set(c, i);
+      occupied[c] = true;
+
+      entry.next_++;
+      entry.count_++;
+      table_index++;
     }
   }
 }
 
-HostConstSharedPtr MaglevTable::chooseHost(uint64_t hash, uint32_t attempt) const {
+MaglevTableSharedPtr
+MaglevTable::createMaglevTable(const NormalizedHostWeightVector& normalized_host_weights,
```

Comment:
It seems a little wonky to me that a base class method should know about subclasses. WDYT about making this a standalone method?